ykcompile/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+        // This is a really dumb register allocator, which runs out of available registers after 7
+        // locals. We can do better than this by using StorageLive/StorageDead from the MIR to free
+        // up registers again, and allocate additional locals on the stack. Though, ultimately we
+        // probably want to implement a proper register allocator, e.g. graph colouring or linear
+        // scan.
+        if l == 0 {
+            0
+        } else {
+            let reg = self
+                .available_regs
+                .pop()
+                .expect("Can't allocate more than 7 locals yet!");
+            *self.assigned_regs.entry(l).or_insert(reg)
+        }
+    }
+
+    /// Move constant `c` of type `usize` into local `a`.
+    pub fn local_usize(&mut self, a: u32, c: usize) {
+        let reg = self.local_to_reg(a);
+        dynasm!(self.asm
+            ; mov Rq(reg), c as i32
+        );
+    }
+
+    /// Move constant `c` of type `u8` into local `a`.
+    pub fn local_u8(&mut self, a: u32, c: u8) {
+        let reg = self.local_to_reg(a);
+        dynasm!(self.asm
+            ; mov Rq(reg), c as i32
+        );
+    }
+
+    /// Move local `var2` into local `var1`.
+    fn local_local(&mut self, var1: u32, var2: u32) {
+        let lreg = self.local_to_reg(var1);
+        let rreg = self.local_to_reg(var2);
+        dynasm!(self.asm
+            ; mov Rq(lreg), Rq(rreg)
+        );
+    }
+
+    fn statement(&mut self, stmt: &Statement) {
+        match stmt {
+            Statement::Assign(l, r) => {
+                let local = l.local.0;
+                match r {
+                    Rvalue::Use(Operand::Place(p)) => self.local_local(local, p.local.0),
+                    Rvalue::Use(Operand::Constant(c)) => match c {
+                        Constant::Int(ci) => match ci {
+                            ConstantInt::UnsignedInt(UnsignedInt::U8(i)) => {
+                                self.local_u8(local, *i)
+                            }
+                            ConstantInt::UnsignedInt(UnsignedInt::Usize(i)) => {
+                                self.local_usize(local, *i)
+                            }
+                            e => panic!("SignedInt, etc: {}", e),
+                        },
+                        Constant::Bool(b) => self.local_u8(local, *b as u8),
+                        c => panic!("Not implemented: {}", c),
+                    },
+                    unimpl => panic!("Not implemented: {:?}", unimpl),
+                };
+            }
+            Statement::Return => {}
+            Statement::Nop => {}
+            Statement::Unimplemented(mir_stmt) => println!("Can't compile: {}", mir_stmt),
+        }
+    }
+
+    fn finish(mut self) -> dynasmrt::ExecutableBuffer {
+        dynasm!(self.asm
+            ; ret
+        );
+        self.asm.finalize().unwrap()
+    }
+
+    pub fn compile(tt: TirTrace) -> CompiledTrace {
+        let regs = vec![11, 10, 9, 8, 2, 1];
```

Comment:
Shalll we comment the mnemonic names of these registers? Is that r11-r8, rbx, rdx?

Reply:
Done.